### PR TITLE
validation, refactor: add missing thread safety lock assertions

### PR DESCRIPTION
**Pull request by jonatack**
A number of functions in validation.{h,cpp} have a thread safety lock annotation in the declaration but are missing the corresponding run-time lock assertion in the definition.

### CODE REVIEW

**Comment by jonatack** [comment]
How to review: verify that a Clang debug build is clean (no -Wthread-safety-analysis warnings) and that each lock assertion (`AssertLockHeld`) has a corresponding lock annotation (`EXCLUSIVE_LOCKS_REQUIRED`) in the declaration.

File: src/validation.cpp [hunk 1]
```diff
@@ -2650,6 +2658,7 @@ bool CChainState::ConnectTip(BlockValidationState& state, CBlockIndex* pindexNew
  * known to be invalid (it's however far from certain to be valid).
  */
 CBlockIndex* CChainState::FindMostWorkChain() {
```
Comment (shaavan) [inline src/validation.cpp:2660]: nit:
```suggestion
CBlockIndex* CChainState::FindMostWorkChain()
{
```

File: src/validation.cpp [hunk 2]
```diff
@@ -1524,6 +1526,8 @@ void CChainState::CheckForkWarningConditions()
 // Called both upon regular invalid block discovery *and* InvalidateBlock
 void CChainState::InvalidChainFound(CBlockIndex* pindexNew)
 {
+    AssertLockHeld(cs_main);
+
```
Comment (shaavan) [inline src/validation.cpp:1530]: nit:

```suggestion
```

File: src/validation.cpp [hunk 3]
```diff
@@ -5014,6 +5031,8 @@ void ChainstateManager::Reset()
 
 void ChainstateManager::MaybeRebalanceCaches()
 {
+    AssertLockHeld(::cs_main);
+
```
Comment (shaavan) [inline src/validation.cpp:5035]: nit:

```suggestion
```

File: src/validation.h [hunk 4]
```diff
@@ -529,6 +529,7 @@ class CChainState
     //! @returns whether or not the CoinsViews object has been fully initialized and we can
     //!          safely flush this object to disk.
     bool CanFlushToDisk() const EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
```
Comment (shaavan) [inline src/validation.h:531]: nit:
```suggestion
    bool CanFlushToDisk() const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
    {
```

**Review by shaavan — COMMENTED** [review]
Concept ACK

I think it makes sense to add the missing thread-safety lock assertions, which prevents the risk of a race condition. 

I shall take a thorough look at this PR before ACKing it. Meanwhile, I found some _nit_ suggestions that should be addressed.

File: src/validation.cpp (hunk 1 above)
Comment (jonatack) [reply src/validation.cpp:2660]: done

File: src/validation.cpp (hunk 2 above)
Comment (jonatack) [reply src/validation.cpp:1530]: done

File: src/validation.cpp (hunk 3 above)
Comment (jonatack) [reply src/validation.cpp:5035]: done

File: src/validation.h (hunk 4 above)
Comment (jonatack) [reply src/validation.h:531]: done

**Comment by jonatack** [comment]
Thanks @shaavan, updated with your feedback.

**Review by hebasto — COMMENTED** [review]
Concept ACK, it follows https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#threads-and-synchronization.

**Review by shaavan — COMMENTED** [review]
I looked into both `src/validation.cpp` and `src/validation.h`, and I found a list of functions that are declared using `EXCLUSIVE_LOCK_REQUIRED()` but does not have `AssertLockHeld()` in their definition, and are also not yet addressed in this PR.

Declared in src/validation.cpp:

- **LimitMempoolSize** `[EXCLUSIVE_LOCKS_REQUIRED(pool.cs, ::cs_main)]`
- **CheckFeeRate** `[EXCLUSIVE_LOCKS_REQUIRED(cs_main, m_pool.cs)]`
- **AcceptToMemoryPool** `[EXCLUSIVE_LOCKS_REQUIRED(cs_main)]`
- **ContextualCheckBlockHeader** `[EXCLUSIVE_LOCKS_REQUIRED(cs_main)]`
- **PreChecks** `[EXCLUSIVE_LOCKS_REQUIRED(cs_main, m_pool.cs)]`
- **PolicyScriptChecks** `[EXCLUSIVE_LOCKS_REQUIRED(cs_main, m_pool.cs)]`
- **ConsensusScriptChecks** `[EXCLUSIVE_LOCKS_REQUIRED(cs_main, m_pool.cs)]`
- **Finalize** `[EXCLUSIVE_LOCKS_REQUIRED(cs_main, m_pool.cs)]`

Declared in src/validation.h:
- ~**DisconnectTip** `[EXCLUSIVE_LOCKS_REQUIRED(cs_main, m_mempool->cs)]` but only `AssertLockHeld(cs_main)`~
- **ProcessTransaction** `[EXCLUSIVE_LOCKS_REQUIRED(cs_main)]`

I think all of these functions come in the scope of this PR, and should be addressed in it. 

Edit: **DisconnectTip()** has `AssertLockHeld()` statement for both the mutex. So no need to make any changes to it. Thanks @jonatack for catching it.
> Done, except for touching the `DisconnectTip()` conditional lock assertion on `m_mempool->cs`

**Comment by jonatack** [comment]
@shaavan thanks! Done, except for touching the `DisconnectTip()` conditional lock assertion on `m_mempool->cs` (same in `ActivateBestChainStep()` and `ConnectTip()`) added in 617661703ac.

**Review by shaavan — APPROVED** [review]
ACK 1c0737e9ccc8f633b4fe06b17d04f83e6d35f4df

Changes since my last review:
- Added missing thread-safety lock for [these](https://github.com/bitcoin/bitcoin/pull/24177#pullrequestreview-867097897) functions.

I was able to successfully compile this PR on Ubuntu 20.04. I agree that this PR can be merged.

**Comment by shaavan** [comment]
Btw I was curious about one thing. Why some functions have `AssertLockHeld(cs_main)` and some `AssertLockHeld(::cs_main)`?

**Comment by jonatack** [comment]
> Btw I was curious about one thing. Why some functions have `AssertLockHeld(cs_main)` and some `AssertLockHeld(::cs_main)`?

Good question. See https://github.com/bitcoin/bitcoin/pull/22932#discussion_r789525111

**Review by shaavan — COMMENTED** [review]
I found a minor nit that I didn't address in my previous review. Feel free to ignore, in case not updating the PR for another reason.

File: src/validation.cpp [hunk 5]
```diff
@@ -288,6 +288,9 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consens
 static void LimitMempoolSize(CTxMemPool& pool, CCoinsViewCache& coins_cache, size_t limit, std::chrono::seconds age)
     EXCLUSIVE_LOCKS_REQUIRED(pool.cs, ::cs_main)
 {
+    AssertLockHeld(pool.cs);
+    AssertLockHeld(::cs_main);
+
```
Comment (shaavan) [inline src/validation.cpp:293]: nit:
```suggestion
```

Comment (jonatack) [reply src/validation.cpp:293]: Ok. Will do if need to retouch.

**Comment by jonatack** [comment]
> each time a function is updated, `cs_main` should be converted to `::cs_main` if not already done

For new code, yes; for updated code, I suppose it depends on the PR author and on reviewer feedback.

(Thanks for reviewing here!)

File: src/validation.cpp (hunk 5 above)
Comment (jonatack) [reply src/validation.cpp:293]: Done

**Comment by jonatack** [comment]
Rebased and updated per `git range-diff 280a777 1c0737e 7ac8b46` to add missing assertions seen in https://github.com/bitcoin/bitcoin/pull/24235#pullrequestreview-873446100 and to replace a lock in `ChainstateManager::Reset()` with thread safety analysis.

**Review by vasild — APPROVED** [review]
ACK 686d35de6bd769efba35177cd9c91f431f46e5a5

`ChainstateManager::Reset()` is called just from the destructor `ChainstateManager::~ChainstateManager()` and from a test, but from the test the destructor is called shortly after. So, maybe the `Reset()` method can be removed, its body copied to the destructor and the explicit call to `Reset()` from the test removed? The only difference is that now, from the test, `Reset()` is called first and `UnloadBlockIndex()` is called afterwards (via the destructor). That order will be swapped, a cursory look shows that should be ok.

**Comment by jonatack** [comment]
Thanks @vasild for having a look.

> maybe the `Reset()` method can be removed, its body copied to the destructor and the explicit call to `Reset()` from the test removed

Done. It looks like `Reset()` was added in 89cdf4d5692 when the requirements probably weren't finalized. Also saw that we can replace the `cs_main` lock in `UnloadBlockIndex()` with a thread safety annotation. Updated per `git diff 686d35d 9a18bdb`.

**Comment by MarcoFalke** [comment]
Mind splitting the inline refactor from the annotations refactor into another pull? If not, that is fine, too.

File: src/validation.cpp [hunk 6]
```diff
@@ -288,6 +288,8 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consens
 static void LimitMempoolSize(CTxMemPool& pool, CCoinsViewCache& coins_cache, size_t limit, std::chrono::seconds age)
     EXCLUSIVE_LOCKS_REQUIRED(pool.cs, ::cs_main)
 {
+    AssertLockHeld(pool.cs);
+    AssertLockHeld(::cs_main);
```
Comment (hebasto) [inline src/validation.cpp:292]: nit: Reorder here and in `EXCLUSIVE_LOCKS_REQUIRED`?

**Review by hebasto — APPROVED** [review]
ACK acdead8a26e92991987efb346566b5474aff8e74, I have reviewed the code (it was a bit tedious :smile:)  and it looks OK, I agree it can be merged.

File: src/validation.cpp (hunk 6 above)
Comment (jonatack) [reply src/validation.cpp]: Done (thanks!)

**Comment by jonatack** [comment]
Updated per `git diff acdead8 f485a07`. @shaavan, @vasild, @hebasto: thank you for reviewing, mind having a final look?

**Review by hebasto — APPROVED** [review]
re-ACK f485a0745455b46390f1c14260643ad69c8fe2ad, only suggested change since my [previous](https://github.com/bitcoin/bitcoin/pull/24177#pullrequestreview-877810465) review.

**Review by vasild — APPROVED** [review]
ACK f485a0745455b46390f1c14260643ad69c8fe2ad

**Pull request merged**